src/osd/OSD.h

```diff
@@ -1827,59 +1869,37 @@ class OSD : public Dispatcher,
     return service.add_map_inc_bl(e, bl);
   }
 
-protected:
+public:
+  // -- shards --
+  vector<OSDShard*> shards;
+  uint32_t num_shards = 0;
+
   // -- placement groups --
-  RWLock pg_map_lock; // this lock orders *above* individual PG _locks
-  ceph::unordered_map<spg_t, PG*> pg_map; // protected by pg_map lock
+  std::atomic<size_t> num_pgs = {0};
 
+protected:
   std::mutex pending_creates_lock;
   using create_from_osd_t = std::pair<pg_t, bool /* is primary*/>;
   std::set<create_from_osd_t> pending_creates_from_osd;
   unsigned pending_creates_from_mon = 0;
 
-  map<spg_t, list<PGPeeringEventRef> > peering_wait_for_split;
   PGRecoveryStats pg_recovery_stats;
 
-  PG   *_lookup_lock_pg_with_map_lock_held(spg_t pgid);
-  PG   *_lookup_lock_pg(spg_t pgid);
+  PGRef _lookup_pg(spg_t pgid);
+  PGRef _lookup_lock_pg(spg_t pgid);
```

Comment:
The commit states "lookup_*pg must return PGRef; Otherwise it is fundamentally unsafe, as the PG might get destroyed out from under us without a reference."

But the two you change here are lock functions that lock the PG before returning it. Doesn't that prevent any kind of deletion happening to them until you drop the lock? If that has changed I think we need to look at our teardown handling across the code base more broadly...

Reply:
Hmm, could probably get away with a bare PG* from _lookup_lock_pg; I think the crash that prompted this was from _lookup_pg racing with the PGRef drop removal.

Reply:
Oh, I misread the unlocked ```_lookup_pg()``` since it replaced ```_lookup_lock_pg_with_map_lock_held()```. Probably best to keep them symmetrical.